FaunaDB.Client/Types/Encoder.cs

```diff
@@ -0,0 +1,278 @@
+                throw new InvalidOperationException($"Self referencing loop detected for object `{obj}`");
+
+            try
+            {
+                stack.Push(obj);
+
+                return EncodeIntern(obj);
+            }
+            finally
+            {
+                stack.Pop();
+            }
+        }
+
+        Value EncodeIntern(object obj)
+        {
+            var type = obj.GetType();
+
+            switch (Type.GetTypeCode(type))
+            {
+                case TypeCode.String:
+                    return StringV.Of((string)obj);
+
+                case TypeCode.Boolean:
+                    return BooleanV.Of((bool)obj);
+
+                case TypeCode.DateTime:
+                    return WrapDateTime((DateTime)obj);
+
+                case TypeCode.SByte:
+                case TypeCode.Int16:
+                case TypeCode.Int32:
+                case TypeCode.Int64:
+                    return LongV.Of(Convert.ToInt64(obj));
+
+                case TypeCode.Char:
+                case TypeCode.Byte:
+                case TypeCode.UInt16:
+                case TypeCode.UInt32:
+                case TypeCode.UInt64:
+                    return LongV.Of(Convert.ToInt64(obj));
+
+                case TypeCode.Single:
+                case TypeCode.Double:
+                case TypeCode.Decimal:
+                    return DoubleV.Of(Convert.ToDouble(obj));
+
+                case TypeCode.Empty:
+                    return NullV.Instance;
+
+                case TypeCode.Object:
+                    if (typeof(byte[]).IsInstanceOfType(obj))
+                        return new BytesV((byte[])obj);
+
+                    if (typeof(IDictionary).IsInstanceOfType(obj))
+                        return WrapDictionary((IDictionary)obj);
+
+                    if (typeof(IEnumerable).IsInstanceOfType(obj))
+                        return WrapEnumerable((IEnumerable)obj);
+
+                    return WrapObj(obj);
+            }
+
+            throw new InvalidOperationException($"Unsupported type {type} at `{obj}`");
+        }
+
+        Value WrapDateTime(DateTime date)
+        {
+            if (date.Ticks % (24 * 60 * 60 * 10000) > 0)
+                return new TimeV(date);
+
+            return new DateV(date);
+        }
+
+        Value WrapEnumerable(IEnumerable array)
+        {
+            var ret = new List<Value>();
+
+            foreach (var value in array)
+                ret.Add(Encode(value));
+
+            return new ArrayV(ret);
+        }
+
+        Value WrapDictionary(IDictionary dic)
+        {
+            var ret = new Dictionary<string, Value>();
+
+            foreach (DictionaryEntry entry in dic)
+                ret.Add(entry.Key.ToString(), Encode(entry.Value));
+
+            return new ObjectV(ret);
+        }
+
+        Value WrapObj(object obj)
+        {
+            var type = obj.GetType();
+
+            Converter converter;
+
+            lock (converters)
```

Comment:
Ok, in my test I ran 100 threads and each thread did 20000 encodings. Using `locks` it ran at `00:00:27.9190770` and using `ReaderWriterLockSlim` it ran at `00:00:26.4105080`.

My conclusion is that it won't make much difference using `lock` compared to `ReaderWriterLockSlim` given the contention is too low and the overhead added by `ReaderWriterLockSlim` is kinda high.

Please have look in my test and see if I'm doing something completely wrong

```csharp
        public void TestThread()
        {
            var bar = new Bar();

            int maxThreads = 100;

            ParameterizedThreadStart thread = delegate(object obj)
            {
                for (int i = 0; i < 2000000 / maxThreads; i++)
                    Encode(bar);
            };

            Thread[] threads = new Thread[maxThreads];
            for (int i = 0; i < maxThreads; i++)
                threads[i] = new Thread(thread);

            var begin = DateTime.Now;

            for (int i = 0; i < maxThreads; i++)
                threads[i].Start();

            for (int i = 0; i < maxThreads; i++)
                threads[i].Join();
            
            var end = DateTime.Now;

            var span = end - begin;
            Console.WriteLine(span);
        }
```